File: zig/src/database/grant.zig
const std = @import("std");

pub const PortGrantRecord = struct {
    id: [36]u8,
    user_id: i32,
    node: []const u8,
    protocol: []const u8,
    host_port: u16,
};

pub const DomainGrantRecord = struct {
    id: [36]u8,
    user_id: i32,
    fqdn: []const u8,
};

File: python/unsafie_cloud/grant.py
from unsafie_cloud.models import GrantResult


class GrantManager:
    def __init__(self, client):
        self.client = client

    async def port(
        self,
        owner: str,
        node: str,
        host_port: int,
        protocol: str = "tcp",
    ) -> GrantResult:
        payload = {
            "owner": owner,
            "node": node,
            "protocol": protocol,
            "host_port": host_port,
        }
        res = await self.client.rpc("admin.grant_port", payload)
        return GrantResult(
            status=res.get("status", "ok"),
            owner=owner,
            granted_resource=f"{node}:{protocol}:{host_port}",
            changed=True,
        )

    async def domain(self, owner: str, fqdn: str) -> GrantResult:
        payload = {
            "owner": owner,
            "fqdn": fqdn,
        }
        res = await self.client.rpc("admin.grant_domain", payload)
        return GrantResult(
            status=res.get("status", "ok"),
            owner=owner,
            granted_resource=fqdn,
            changed=True,
        )
